src/node_buffer.cc

```diff
@@ -602,6 +602,110 @@ void Compare(const FunctionCallbackInfo<Value> &args) {
 }
 
 
+int32_t IndexOf(char* haystack, size_t h_len, char* needle, size_t n_len) {
+  CHECK_GE(h_len, n_len);
+  int32_t ret = -1;
+  size_t i = 0;
+  // TODO(trevnorris): Implement Boyer-Moore string search algorithm.
+  for (; i < h_len - n_len; i++) {
+    if (haystack[i] == needle[0]) {
+      ret = memcmp(haystack + i, needle, n_len);
+      if (ret == 0)
+        break;
+    }
+  }
+  return ret == 0 ? i : -1;
+}
+
+
+void IndexOfString(const FunctionCallbackInfo<Value>& args) {
+  ASSERT(args[0]->IsObject());
+  ASSERT(args[1]->IsString());
+  ASSERT(args[2]->IsNumber());
+
+  ARGS_THIS(args[0].As<Object>());
+  node::Utf8Value str(args.GetIsolate(), args[1]);
+  int32_t offset = args[2]->Int32Value();
+
+  if (offset < 0) {
+    if (offset + static_cast<int32_t>(obj_length) < 0)
+      offset = 0;
+    else
+      offset = obj_length + offset;
+  }
+
+  if (str.length() == 0 ||
+      obj_length == 0 ||
+      (str.length() + offset <= str.length() && offset != 0) ||
+      str.length() + offset > obj_length)
+    return args.GetReturnValue().Set(-1);
+
+  int32_t r =
+      IndexOf(obj_data + offset, obj_length - offset, *str, str.length());
+  args.GetReturnValue().Set(r == -1 ? -1 : r + offset);
+}
+
+
+void IndexOfBuffer(const FunctionCallbackInfo<Value>& args) {
+  ASSERT(args[0]->IsObject());
+  ASSERT(args[1]->IsObject());
+  ASSERT(args[2]->IsNumber());
+
+  ARGS_THIS(args[0].As<Object>());
+  Local<Object> buf = args[1].As<Object>();
+  int32_t offset = args[2]->Int32Value();
+  size_t buf_length = buf->GetIndexedPropertiesExternalArrayDataLength();
+  char* buf_data =
+      static_cast<char*>(buf->GetIndexedPropertiesExternalArrayData());
+  if (buf_length > 0)
+    CHECK_NE(buf_data, nullptr);
+
+  if (offset < 0) {
+    if (offset + static_cast<int32_t>(obj_length) < 0)
+      offset = 0;
+    else
+      offset = obj_length + offset;
+  }
+
+  if (buf_length == 0 ||
+      obj_length == 0 ||
+      (buf_length + offset <= buf_length && offset != 0) ||
+      buf_length + offset > obj_length)
+    return args.GetReturnValue().Set(-1);
+
+  int32_t r =
+    IndexOf(obj_data + offset, obj_length - offset, buf_data, buf_length);
+  args.GetReturnValue().Set(r == -1 ? -1 : r + offset);
+}
+
+
+void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+  ASSERT(args[0]->IsObject());
+  ASSERT(args[1]->IsNumber());
+  ASSERT(args[2]->IsNumber());
+
+  ARGS_THIS(args[0].As<Object>());
+  int32_t needle = args[1]->Int32Value();
```

Comment:
documentation states the value is a uint8, can range from 0 to 255. i'll make the change here to reflect that.